enhancements/ota/distribute-secondary-metadata-as-container-image.md

```diff
@@ -0,0 +1,165 @@
+
+### Container image content integrity
+The content of the cincinnati-graph-data repository may be altered by the process of being packaged by CPaaS and distributed via the container image registry.
+
+#### Mitigation
+A mitigation strategy is to sign the resulting image in the CPaaS pipeline, publish the signatures, and have Cincinnati verify the image after download.
+
+### cincinnati-graph-data source repository integrity
+Not strictly related to this enhancement, is the fact that we are not signing the commits in the source repository in the first place.
+By this we effectively trust GitHub not to change the content.
+
+#### Mitigation
+Enforce signed commits in the source repository.
+
+## Design Details
+
+### Test Plan
+**Note:** *Section not required until targeted at a release.*
+
+Consider the following in developing a test plan for this enhancement:
+- Will there be e2e and integration tests, in addition to unit tests?
+- How will it be tested in isolation vs with other components?
+
+No need to outline all of the test cases, just the general strategy. Anything
+that would count as tricky in the implementation and anything particularly
+challenging to test should be called out.
+
+All code is expected to have adequate tests (eventually with coverage
+expectations).
+
+### Graduation Criteria
+**Note:** *Section not required until targeted at a release.*
+
+- Maturity levels - `Developer Preview`, `Red Hat Staging`, `Red Hat Production`, `GA`
+
+##### `Developer Preview`
+- New Cincinnati plugin
+- Container image registry managed by and dedicated to prow CI
+- Integration tests using an image (to-be-determined: verifying against mocked signatures)
+- CPaaS pipeline is setup up to
+
+##### `Developer Preview` -> `Red Hat Staging`
+- Update the Cincinnati deployment template to use the new plugin.
+
+##### Red Hat Staging -> Red Hat Production
+- New plugin is successfully deployed to production
+- Adjust the Cincinnati Operator to use the new plugin
+
+##### Red Hat Production -> GA
+- Document the mirroring process for end-users
+- Definitive positive feedback from the ACM team and their use-case
+
+##### Removing a deprecated feature
+*Not applicable*
+
+### Upgrade / Downgrade Strategy
+*Not applicable*
+
+### Version Skew Strategy
+*Not applicable*
+
+## Implementation History
+
+### Initial implementation
+The initial implementation involves setting up the CPaaS pipeline, implementing a new Cincinnati plugin which downloads the published image and extracts it to disk for subsequent processing in the plugin chain.
+It also involves adjusting the Cincinnati Operator to use the new plugin by default.
+
+## Drawbacks
+
+### Increase OCP release latency
+The enhancement introduces the additional latency of the CPaaS pipeline into the OCP release workflow.
+Assuming that the pipeline is fully automated, the latency introduced is expected to be several orders of magnitude lower (minutes) compared to the OCP release cadence (hours-days).
```

Comment:
My concern is about edge-pulling, where our latency is ideally a small fraction of our future phased-rollout duration.  The failure detection timeline (in an ideal, future world):

1. Update becomes available to a cluster.
2. Cluster updates (could be ~minutes with a future auto-update knob, #124, could be hours/days/never with manual updates).
3. Cluster begins to hit update-triggered issues (sometime between 0 and 3h, with a cap of 3h for "update taking surprisingly long counting as a "failure" and contributing to inspection).
4. Cluster pushes Telemetry up describing it's issue (pushes every ~5m).
5. Alerting trips over the concerning Telemetry and pings for review, while also automatically PRing and merging graph-data to remove the edge for safety during triage (~minutes).
6. CPaaS informed of the merge, builds the image, pushes to Quay (or wherever).
7. Cincinnati (operator?) notices the new image, verifies the sig, pull it down, and starts serving it (~minutes).
8. Rest of fleet pings Cincinnati, notices that the edge is gone (~minutes).

There's an unavoidable ~hours delay in step 3, and smearing out the roll-out minimizes the number of clusters that initiate the update during that window.  If step 6 is also ~hours, we'd need to ~double our roll-out durations to maintain the same level of protection.  If step 6 is minutes, the duration would only have to increase slightly to preserve the same level of protection.

Reply:
>  If step 6 is minutes, the duration would only have to increase slightly to preserve the same level of protection.

I assume it's minutes as the build job doesn't rely on a huge toolchain. The steps are: check out the git repo, build a container image with the files, sign the image, and publish the image and the signatures.

Reply:
Minutes sounds reasonable if all those steps are automated, but it's not obvious to me that they are (signing might require a manual step?).  There's also whatever logic that is informing CPaaS of the merge, since my impression (which may be wrong) is that CPaaS can't listen to GitHub merge notifications.  But regardless, "the OCP release cadence" is not what the latency section should use as a benchmark; it should be talking about edge blocking.  If you can show CPaaS builds with ~minutes latencies, that's great.  If you can wave your hands around "should be ~minutes", that's probably fine too.  But folks shouldn't have to drop into PR comments to find these arguments ;).

Reply:
@LalatenduMohanty what's your impression on the CPaaS fully-automated build? Mine is that this is possible today.

Reply:
@LalatenduMohanty :ping_pong:

Reply:
It is not the CPaaS build I am worried about, but the whole process of signing the image, creating an errata and pushing errata to live would take hours for sure. If I am proven wrong I would be happy about it.

Reply:
Here's an excerpt from a thread on the CPaaS Google Chat:

```
Eyal Edri,
Jul 17, 4:24 PM
,
@Stefan Junker loos at this run for e.g https://jenkins-cpaas-cnv.cloud.paas.psi.redhat.com/blue/organizations/jenkins/cnv-2.4.0%2FBuild-Pipeline/detail/Build-Pipeline/67/pipeline
it built multiple containers, RPMs, operator, attached to Errata and run smoke tests, in 1.3 hours
```

All of this happens manually, and our builds are likely faster than the CNV build as we don't invoke any toolchains and also don't build RPMs.

Reply:
Here's an excerpt from a thread on the CPaaS Google Chat:

```
Eyal Edri,
Jul 17, 4:24 PM
,
@Stefan Junker loos at this run for e.g https://jenkins-cpaas-cnv.cloud.paas.psi.redhat.com/blue/organizations/jenkins/cnv-2.4.0%2FBuild-Pipeline/detail/Build-Pipeline/67/pipeline
it built multiple containers, RPMs, operator, attached to Errata and run smoke tests, in 1.3 hours
```

All of this happens automatically, and our builds are likely faster than the CNV build as we don't invoke any toolchains and also don't build RPMs.